File: runtimes/swiftui/Sources/StyleConverterRuntime/StyleEngine/borders/image/BorderImageExtractor.swift
//
//  BorderImageExtractor.swift
//  StyleEngine/borders/image — Phase 5.
//
//  Pulls the five border-image properties out of the IR:
//    BorderImageSource / Slice / Width / Outset / Repeat.
//
//  IR shapes (see parser folder
//  `src/main/kotlin/app/parsing/css/properties/longhands/borders/image/`):
//    Source : {"type":"url","url":"..."} | {"type":"gradient","gradient":"..."}
//             | {"type":"none"}
//    Slice  : {"top": {"type":"number","value":N}, ..., "fill": true|false}
//    Width  : {"top": <dimension>, ..., "left": <dimension>}
//             dimension = {"type":"length","px":N} | {"type":"percentage","value":N}
//                       | {"type":"number","value":N} | {"type":"auto"}
//    Outset : same shape as width.
//    Repeat : "STRETCH" | {"horizontal":"REPEAT","vertical":"STRETCH"}
//
//  Mirrors the Android extractor byte-for-byte.
//

import Foundation
import CoreGraphics

enum BorderImageExtractor {

    // Register these in PropertyRegistry.migrated to skip the legacy switch.
    static let propertyNames: [String] = [
        "BorderImageSource", "BorderImageSlice", "BorderImageWidth",
        "BorderImageOutset", "BorderImageRepeat",
    ]

    // Returns nil when none of the five keys appear so the applier can
    // short-circuit.
    static func extract(from properties: [IRProperty]) -> BorderImageConfig? {
        var cfg = BorderImageConfig()
        var touched = false

        for prop in properties {
            switch prop.type {
            case "BorderImageSource":
                cfg.source = extractSource(prop.data); touched = true
            case "BorderImageSlice":
                extractSlice(prop.data, into: &cfg); touched = true
            case "BorderImageWidth":
                let t = extractFourValues(prop.data)
                cfg.widthTop = t.top; cfg.widthRight = t.right
                cfg.widthBottom = t.bottom; cfg.widthLeft = t.left
                touched = true
            case "BorderImageOutset":
                let t = extractFourValues(prop.data)
                cfg.outsetTop = t.top; cfg.outsetRight = t.right
                cfg.outsetBottom = t.bottom; cfg.outsetLeft = t.left
                touched = true
            case "BorderImageRepeat":
                let r = extractRepeat(prop.data)
                cfg.repeatHorizontal = r.h; cfg.repeatVertical = r.v
                touched = true
            default: break
            }
        }
        return touched ? cfg : nil
    }

    // MARK: - Source

    // Dispatch on the `type` tag. Unknown tags → .none so CSS's
    // "invalid values reset the property" rule holds.
    private static func extractSource(_ v: IRValue?) -> BorderImageSource {
        guard case .object(let o) = v else { return .none }
        let t = o["type"]?.stringValue?.lowercased() ?? ""
        switch t {
        case "url":      return .url(o["url"]?.stringValue ?? "")
        case "gradient": return .gradient(o["gradient"]?.stringValue ?? "")
        default:         return .none
        }
    }

    // MARK: - Slice

    // Populates the four slice edges + fill in-place. Separate function
    // so the main loop stays flat.
    private static func extractSlice(_ v: IRValue?,
                                     into cfg: inout BorderImageConfig) {
        guard case .object(let o) = v else { return }
        cfg.sliceTop    = extractSliceEdge(o["top"])
        cfg.sliceRight  = extractSliceEdge(o["right"])
        cfg.sliceBottom = extractSliceEdge(o["bottom"])
        cfg.sliceLeft   = extractSliceEdge(o["left"])
        // `fill` arrives either as a real boolean or as the string "true"
        // / "fill" depending on parser flavour — cover both paths.
        if let b = o["fill"]?.boolValue {
            cfg.sliceFill = b
        } else if let s = o["fill"]?.stringValue?.lowercased() {
            cfg.sliceFill = (s == "true" || s == "fill")
        }
    }

    // Edge shape: `{type: "number"|"percentage", value: N}`.
    private static func extractSliceEdge(_ v: IRValue?) -> BorderImageSliceEdge? {
        guard case .object(let o) = v,
              let t = o["type"]?.stringValue?.lowercased() else { return nil }
        // Try both `value` and `percentage` — different parser versions
        // name the scalar differently.
        let raw = (o["value"] ?? o["percentage"])?.doubleValue
        switch t {
        case "number":
            guard let n = raw else { return nil }
            return BorderImageSliceEdge(value: CGFloat(n), isPercent: false)
        case "percentage":
            guard let n = raw else { return nil }
            return BorderImageSliceEdge(value: CGFloat(n), isPercent: true)
        default:
            return nil
        }
    }

    // MARK: - Width / Outset (shared four-dimension shape)

    // Tuple-return to keep the caller readable.
    private struct FourDims {
        var top: BorderImageDimension?
        var right: BorderImageDimension?
        var bottom: BorderImageDimension?
        var left: BorderImageDimension?
    }

    private static func extractFourValues(_ v: IRValue?) -> FourDims {
        guard case .object(let o) = v else { return FourDims() }
        return FourDims(
            top:    extractDim(o["top"]),
            right:  extractDim(o["right"]),
            bottom: extractDim(o["bottom"]),
            left:   extractDim(o["left"])
        )
    }

    private static func extractDim(_ v: IRValue?) -> BorderImageDimension? {
        guard case .object(let o) = v,
              let t = o["type"]?.stringValue?.lowercased() else { return nil }
        switch t {
        case "auto":   return .auto
        case "length":
            // Reuse the Phase 1 px extractor — handles `{px: N}` + wrappers.
            guard let px = ValueExtractors.extractPx(.object(o)) else { return nil }
            return .length(px)
        case "percentage":
            let n = (o["value"] ?? o["percentage"])?.doubleValue ?? 0
            return .percent(CGFloat(n))
        case "number":
            guard let n = o["value"]?.doubleValue else { return nil }
            return .number(CGFloat(n))
        default:
            return nil
        }
    }

    // MARK: - Repeat

    // Two-axis result. Name-dodging the Swift `Repeat` builtin type.
    private struct RepeatAxes { var h: BorderImageRepeat; var v: BorderImageRepeat }

    private static func extractRepeat(_ v: IRValue?) -> RepeatAxes {
        // Single-string form applies to both axes.
        if case .string(let s) = v, let k = keyword(s) {
            return RepeatAxes(h: k, v: k)
        }
        // Object form separates horizontal/vertical.
        if case .object(let o) = v {
            let h = keyword(o["horizontal"]?.stringValue) ?? .stretch
            let vt = keyword(o["vertical"]?.stringValue) ?? h
            return RepeatAxes(h: h, v: vt)
        }
        return RepeatAxes(h: .stretch, v: .stretch)
    }

    // Upper-case keyword → enum. Unknowns collapse to .stretch (CSS default).
    private static func keyword(_ raw: String?) -> BorderImageRepeat? {
        guard let raw = raw?.uppercased() else { return nil }
        switch raw {
        case "STRETCH": return .stretch
        case "REPEAT":  return .repeatTile
        case "ROUND":   return .round
        case "SPACE":   return .space
        default:        return nil
        }
    }
}

File: testing/Android/app/src/main/java/com/styleconverter/test/style/borders/image/BorderImageExtractor.kt
package com.styleconverter.test.style.borders.image

import com.styleconverter.test.style.PropertyRegistry
import com.styleconverter.test.style.core.types.ValueExtractors
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.floatOrNull
import kotlinx.serialization.json.jsonPrimitive

/**
 * Extracts border-image configuration from IR properties.
 */
object BorderImageExtractor {

    init {
        // The 5 CSS border-image longhands (source / slice / width / outset /
        // repeat). The applier is a Composable (BorderImageBox) rather than
        // a Modifier, so the legacy dispatch still has to route to it — we
        // register them here only so the coverage report knows they're owned.
        PropertyRegistry.migrated(
            "BorderImageSource", "BorderImageSlice", "BorderImageWidth",
            "BorderImageOutset", "BorderImageRepeat",
            owner = "borders/image"
        )
    }

    /**
     * Extract complete border-image configuration from property pairs.
     */
    fun extractBorderImageConfig(properties: List<Pair<String, JsonElement?>>): BorderImageConfig {
        var source: BorderImageSourceValue = BorderImageSourceValue.None
        var sliceTop: BorderImageSliceEdge? = null
        var sliceRight: BorderImageSliceEdge? = null
        var sliceBottom: BorderImageSliceEdge? = null
        var sliceLeft: BorderImageSliceEdge? = null
        var sliceFill = false
        var widthTop: BorderImageDimension? = null
        var widthRight: BorderImageDimension? = null
        var widthBottom: BorderImageDimension? = null
        var widthLeft: BorderImageDimension? = null
        var outsetTop: BorderImageDimension? = null
        var outsetRight: BorderImageDimension? = null
        var outsetBottom: BorderImageDimension? = null
        var outsetLeft: BorderImageDimension? = null
        var repeatHorizontal = BorderImageRepeatValue.STRETCH
        var repeatVertical = BorderImageRepeatValue.STRETCH

        for ((type, data) in properties) {
            when (type) {
                "BorderImageSource" -> source = extractBorderImageSource(data)
                "BorderImageSlice" -> {
                    val result = extractBorderImageSlice(data)
                    sliceTop = result.top
                    sliceRight = result.right
                    sliceBottom = result.bottom
                    sliceLeft = result.left
                    sliceFill = result.fill
                }
                "BorderImageWidth" -> {
                    val result = extractBorderImageFourValues(data)
                    widthTop = result.top
                    widthRight = result.right
                    widthBottom = result.bottom
                    widthLeft = result.left
                }
                "BorderImageOutset" -> {
                    val result = extractBorderImageFourValues(data)
                    outsetTop = result.top
                    outsetRight = result.right
                    outsetBottom = result.bottom
                    outsetLeft = result.left
                }
                "BorderImageRepeat" -> {
                    val result = extractBorderImageRepeat(data)
                    repeatHorizontal = result.first
                    repeatVertical = result.second
                }
            }
        }

        return BorderImageConfig(
            source = source,
            sliceTop = sliceTop,
            sliceRight = sliceRight,
            sliceBottom = sliceBottom,
            sliceLeft = sliceLeft,
            sliceFill = sliceFill,
            widthTop = widthTop,
            widthRight = widthRight,
            widthBottom = widthBottom,
            widthLeft = widthLeft,
            outsetTop = outsetTop,
            outsetRight = outsetRight,
            outsetBottom = outsetBottom,
            outsetLeft = outsetLeft,
            repeatHorizontal = repeatHorizontal,
            repeatVertical = repeatVertical
        )
    }

    /**
     * Extract border-image-source from IR data.
     */
    fun extractBorderImageSource(data: JsonElement?): BorderImageSourceValue {
        if (data == null) return BorderImageSourceValue.None
        if (data !is JsonObject) return BorderImageSourceValue.None

        val sourceData = (data["source"] as? JsonObject) ?: data
        val type = sourceData["type"]?.jsonPrimitive?.contentOrNull

        return when (type?.lowercase()) {
            "url" -> {
                val url = sourceData["url"]?.jsonPrimitive?.contentOrNull ?: ""
                BorderImageSourceValue.Url(url)
            }
            "gradient" -> {
                val gradient = sourceData["gradient"]?.jsonPrimitive?.contentOrNull ?: ""
                BorderImageSourceValue.Gradient(gradient)
            }
            "none" -> BorderImageSourceValue.None
            else -> BorderImageSourceValue.None
        }
    }

    /**
     * Extract border-image-slice values.
     */
    private fun extractBorderImageSlice(data: JsonElement?): SliceResult {
        if (data == null) return SliceResult()
        if (data !is JsonObject) return SliceResult()

        val top = extractSliceEdge(data["top"])
        val right = extractSliceEdge(data["right"])
        val bottom = extractSliceEdge(data["bottom"])
        val left = extractSliceEdge(data["left"])
        val fill = data["fill"]?.jsonPrimitive?.contentOrNull?.lowercase() == "true" ||
                   data["fill"]?.jsonPrimitive?.contentOrNull?.lowercase() == "fill"

        return SliceResult(top, right, bottom, left, fill)
    }

    private fun extractSliceEdge(data: JsonElement?): BorderImageSliceEdge? {
        if (data == null) return null
        if (data !is JsonObject) return null

        val type = data["type"]?.jsonPrimitive?.contentOrNull
        return when (type?.lowercase()) {
            "number" -> {
                val value = data["value"]?.jsonPrimitive?.floatOrNull
                    ?: (data["value"] as? JsonObject)?.get("value")?.jsonPrimitive?.floatOrNull
                    ?: return null
                BorderImageSliceEdge(value, isPercentage = false)
            }
            "percentage" -> {
                val value = data["percentage"]?.jsonPrimitive?.floatOrNull
                    ?: data["value"]?.jsonPrimitive?.floatOrNull
                    ?: return null
                BorderImageSliceEdge(value, isPercentage = true)
            }
            else -> null
        }
    }

    /**
     * Extract four-value properties (width, outset).
     */
    private fun extractBorderImageFourValues(data: JsonElement?): FourValuesResult {
        if (data == null) return FourValuesResult()
        if (data !is JsonObject) return FourValuesResult()

        val top = extractDimension(data["top"])
        val right = extractDimension(data["right"])
        val bottom = extractDimension(data["bottom"])
        val left = extractDimension(data["left"])

        return FourValuesResult(top, right, bottom, left)
    }

    private fun extractDimension(data: JsonElement?): BorderImageDimension? {
        if (data == null) return null
        if (data !is JsonObject) return null

        val type = data["type"]?.jsonPrimitive?.contentOrNull
        return when (type?.lowercase()) {
            "auto" -> BorderImageDimension.Auto
            "length" -> {
                val dp = ValueExtractors.extractDp(data)
                if (dp != null) BorderImageDimension.Length(dp) else null
            }
            "percentage" -> {
                val value = data["percentage"]?.jsonPrimitive?.floatOrNull
                    ?: data["value"]?.jsonPrimitive?.floatOrNull
                    ?: return null
                BorderImageDimension.Percentage(value)
            }
            "number" -> {
                val value = data["value"]?.jsonPrimitive?.floatOrNull ?: return null
                BorderImageDimension.Number(value)
            }
            else -> null
        }
    }

    /**
     * Extract border-image-repeat values.
     */
    private fun extractBorderImageRepeat(data: JsonElement?): Pair<BorderImageRepeatValue, BorderImageRepeatValue> {
        if (data == null) return Pair(BorderImageRepeatValue.STRETCH, BorderImageRepeatValue.STRETCH)
        if (data !is JsonObject) return Pair(BorderImageRepeatValue.STRETCH, BorderImageRepeatValue.STRETCH)

        val horizontal = extractRepeatValue(data["horizontal"]?.jsonPrimitive?.contentOrNull)
        val vertical = extractRepeatValue(data["vertical"]?.jsonPrimitive?.contentOrNull)
            ?: horizontal

        return Pair(horizontal, vertical)
    }

    private fun extractRepeatValue(keyword: String?): BorderImageRepeatValue {
        return when (keyword?.uppercase()) {
            "STRETCH" -> BorderImageRepeatValue.STRETCH
            "REPEAT" -> BorderImageRepeatValue.REPEAT
            "ROUND" -> BorderImageRepeatValue.ROUND
            "SPACE" -> BorderImageRepeatValue.SPACE
            else -> BorderImageRepeatValue.STRETCH
        }
    }

    /**
     * Check if a property type is border-image related.
     */
    fun isBorderImageProperty(type: String): Boolean {
        return type in BORDER_IMAGE_PROPERTIES
    }

    private val BORDER_IMAGE_PROPERTIES = setOf(
        "BorderImageSource", "BorderImageSlice", "BorderImageWidth",
        "BorderImageOutset", "BorderImageRepeat"
    )

    // Helper data classes
    private data class SliceResult(
        val top: BorderImageSliceEdge? = null,
        val right: BorderImageSliceEdge? = null,
        val bottom: BorderImageSliceEdge? = null,
        val left: BorderImageSliceEdge? = null,
        val fill: Boolean = false
    )

    private data class FourValuesResult(
        val top: BorderImageDimension? = null,
        val right: BorderImageDimension? = null,
        val bottom: BorderImageDimension? = null,
        val left: BorderImageDimension? = null
    )
}
